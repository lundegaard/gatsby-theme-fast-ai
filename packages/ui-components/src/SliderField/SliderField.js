import React, { Children, Fragment } from 'react';
import PropTypes from 'prop-types';
import { compose, map } from 'ramda';
import { ensureArray } from 'ramda-extension';

import useGeneratedId from '../hooks/useGeneratedId';
import useDebounce from '../hooks/useDebounce';
import Slider from '../Slider';
import Text from '../Text';
import Box from '../Box';
import FormGroup, { getVariant } from '../FormGroup';

const DefaultRenderValues = ({ children }) =>
	Children.map(children, (child, i) => (
		<Fragment>
			{i === 0 ? '' : ' - '}
			{child}
		</Fragment>
	));

const SliderField = ({
	id: idProp,
	renderValue: Value,
	renderValues: RenderValues = DefaultRenderValues,
	label,
	hasError,
	disabled,
	readOnly,
	hint,
	value,
	...rest
}) => {
	const generatedStub = useGeneratedId();
	const generatedId = `field-${generatedStub}`;
	const id = idProp || generatedId;
	const variant = getVariant({ disabled, readOnly, hasError });
	const [valueDebounced] = useDebounce(value, 200);

	const getValue = (x) => (Value ? <Value key={x}>{x}</Value> : x);
	return (
		<Box sx={{ position: 'relative' }}>
			<Text
				sx={{
					textAlign: 'right',
					position: ['absolute', 'absolute', 'static'],
					right: 0,
					top: 0,
				}}
				mb={0}
			>
				<RenderValues>
					{compose(map(getValue), ensureArray)(valueDebounced)}
				</RenderValues>
			</Text>
			<FormGroup
				id={id}
				label={label}
				hint={hint}
				disabled={disabled}
				readOnly={readOnly}
				hasError={hasError}
			>
				<Slider
					variant={variant}
					disabled={disabled}
					readOnly={readOnly}
					id={id}
					value={value}
					{...rest}
				/>
			</FormGroup>
		</Box>
	);
};

SliderField.propTypes = {
	/** Disabled state */
	disabled: PropTypes.bool,
	/** Error state */
	hasError: PropTypes.bool,
	/** Hint itext */
	hint: PropTypes.node,
	/** Used to connect label and slider. If not defined, component will generate
	 * its own. */
	id: PropTypes.string,
	/** Label for Slider  */
	label: PropTypes.node,
	/** Called when `onChange` event is triggered. */
	onChange: PropTypes.func,
	/** Read-only state */
	readOnly: PropTypes.bool,
	/** Formattting component for the value */
	renderValue: PropTypes.elementType,
	/** Formattting for array of values */
	renderValues: PropTypes.func,
	/** Current value of checkbox. */
	value: PropTypes.any,
};

export default SliderField;
