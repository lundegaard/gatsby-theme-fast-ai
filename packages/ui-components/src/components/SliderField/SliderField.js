import React from 'react';
import PropTypes from 'prop-types';

import useGeneratedId from '../../hooks/useGeneratedId';
import Label from '../Label';
import Slider from '../Slider';
import Box from '../Box';
import Text from '../Text';
import SuperFieldHint from '../SuperFieldHint';

const getVariant = ({ hasError, readOnly, disabled }) => {
	if (hasError) {
		return 'error';
	}
	if (disabled || readOnly) {
		return 'disabled';
	}
};

const SliderField = ({
	id: idProp,
	renderValue: Value,
	label,
	hasError,
	disabled,
	readOnly,
	hint,
	value,
	...rest
}) => {
	const generatedStub = useGeneratedId();
	const generatedId = `field-${generatedStub}`;
	const id = idProp || generatedId;
	const variant = getVariant({ hasError, readOnly, disabled });
	return (
		<Box sx={{ position: 'relative' }} variant={variant} pb={2}>
			<Label htmlFor={id}>{label}</Label>
			<Text
				variant={variant ? variant : 'secondary'}
				sx={{ position: 'absolute', color: variant ? null : 'secondary', right: 0, top: 0 }}
			>
				<Value>{value}</Value>
			</Text>
			<Slider
				variant={variant ? `forms.slider.${variant}` : 'forms.slider.default'}
				disabled={disabled}
				readOnly={readOnly}
				id={id}
				value={value}
				{...rest}
			/>
			{hint ? <SuperFieldHint>{hint}</SuperFieldHint> : null}
		</Box>
	);
};

SliderField.propTypes = {
	/** Disabled state */
	disabled: PropTypes.bool,
	/** Error state */
	hasError: PropTypes.bool,
	/** Hint itext */
	hint: PropTypes.node,
	/** Used to connect label and slider. If not defined, component will generate its own. */
	id: PropTypes.string,
	/** Label for Slider  */
	label: PropTypes.node,
	/** Called when `onChange` event is triggered. */
	onChange: PropTypes.func,
	/** Read-only state */
	readOnly: PropTypes.bool,
	/** Formattting component for the value */
	renderValue: PropTypes.elementType,
	/** Current value of checkbox. */
	value: PropTypes.any,
};

export default SliderField;
