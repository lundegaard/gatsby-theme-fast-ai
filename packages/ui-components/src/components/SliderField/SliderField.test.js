import React from 'react';
import { shallow } from 'enzyme';

import SliderField from './SliderField';

describe('<SliderField />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<SliderField {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
