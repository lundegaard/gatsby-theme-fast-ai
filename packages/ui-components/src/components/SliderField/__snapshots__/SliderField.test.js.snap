// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SliderField /> matches snapshot when shallowly rendered 1`] = `
<Box
  pb={2}
  sx={
    Object {
      "position": "relative",
    }
  }
>
  <Label
    htmlFor="field-0"
  />
  <Text
    sx={
      Object {
        "color": "secondary",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    variant="secondary"
  >
    <Component />
  </Text>
  <Slider
    id="field-0"
    variant="forms.slider.default"
  />
</Box>
`;
